### DISCUSSION TITLE
Questions related to hyperparameter optimization

### DISCUSSION THREAD

**Discussion by ZiJing-Lu** [chemprop/chemprop#1049] [Q&A] [answered]
Hello,
I would like to know how to write Python code in Jupyter Notebook to optimize all the hyperparameters of a model. However, I only found commands related to hyperparameter optimization in the "Command Line Tutorials" section and did not find any implementation using Python code.

Additionally, I found the following statement in the "Data splitting" section under "Python Module Tutorials": "A typical Chemprop workflow uses three sets of data. The first is used to train the model. The second is used as validation for early stopping and hyperparameter optimization. The third is used to test the final model’s performance as an estimate for how it will perform on future data." 
![微信截图_28](https://github.com/user-attachments/assets/cf9e4af5-49c5-4946-bb1a-ebccd325dd7c)

So, how is the hyperparameter optimization process specifically implemented on the validation dataset?

If you have any suggestions for hyperparameter optimization, I would greatly appreciate it.

Thank you

Best Regards
DengFeng

**Comment by ZiJing-Lu** [comment]
Specifically, I would like to know how to optimize these hyperparameters of the model.

![微信截图_2](https://github.com/user-attachments/assets/0f1ed7d9-d1f9-481c-a5d0-44f630ad0af7)

**Comment by JacksonBurns** [comment]
This demo notebook shows how to perform hyper-parameter optimization via code, and how to access the various hyper-parameters of the model and optimize them: https://github.com/chemprop/chemprop/blob/main/examples/hpopting.ipynb

The validation set is used for early stopping during training, and then the best performing set of hyper-parameters based on the validation set is returned as the 'optimal' hyper-parameters.

@KnathanM it looks like in the CLI code for hpopt we never actually use the test data and report the performance on it, as promised in the docs? See: https://github.com/chemprop/chemprop/blob/main/chemprop/cli/hpopt.py#L472C27-L472C36

**Reply by ZiJing-Lu** [reply]
Thank you for your help and suggestions! 

Following your recommendation, I have studied  [https://github.com/chemprop/chemprop/blob/main/examples/hpopting.ipynb](https://github.com/chemprop/chemprop/blob/main/examples/hpopting.ipynb ) and try to apply it to my own multi-component tasks. But there seem to be some unexpected errors. I've tried to debug but still unable to solve these problems yet. Could you please offer some possible solutions?  
#1052

**Reply by JacksonBurns** [reply, marked as answer]
This was resolved in #1052

**Reply by KnathanM** [reply]
@JacksonBurns Can you comment a reference to where the docs it says the test data is used by hpopt? I think our intention was to not use the test set in `chemprop hpopt` so that a user could use `hpopt` and `train` on the same dataset without running into a data leak. (The splitting is the same for both runs, or at least should be.)

**Reply by JacksonBurns** [reply]
It actually doesn't say it's used by `hpopt`, the OP was pulling form this part of the docs which is not part of `hptop`: https://chemprop.readthedocs.io/en/latest/tutorial/python/data/splitting.html#Data-splits - just a misunderstanding
